### PR TITLE
Implement SetThreadPriority

### LINKED ISSUE #784: Implement SetThreadPriority on Unix
I'm tired of looking at the warning in PriorityBoost.cpp, it's about time we got around to implementing this.

It's likely all we need to do is to call the `nice` function or near equivalent, we just need to make sure we do so in a way that will also work properly on Mac.

### PR DESCRIPTION
**Pull request by codemercenary**
@gittyupagain implemented this functionality on a private branch awhile back, let's get it into mainline.

Fixes #784

### CODE REVIEW

File: src/autowiring/CoreThreadLinux.cpp [hunk 1]
```diff
@@ -23,3 +24,39 @@ void BasicThread::GetThreadTimes(std::chrono::milliseconds& kernelTime, std::chr
...
+  case ThreadPriority::Highest:
+    percent = 83;
+    break;
+  case ThreadPriority::TimeCritical:
+    percent = 100;
+    break;
+  default:
+    throw std::runtime_error("Attempted to assign an unrecognized thread priority");
+  }
+  param.sched_priority = PTHREAD_MIN_PRIORITY + (percent*(PTHREAD_MAX_PRIORITY - PTHREAD_MIN_PRIORITY) + 50) / 100;
```
Comment (hham) [inline src/autowiring/CoreThreadLinux.cpp]: #include <sched.h>
int sched_get_priority_max(int policy);
int sched_get_priority_min(int policy);
should be used here.

File: src/autowiring/CoreThreadMac.cpp [hunk 2]
```diff
@@ -42,3 +51,38 @@ void BasicThread::GetThreadTimes(std::chrono::milliseconds& kernelTime, std::chr
...
+    break;
+  case ThreadPriority::TimeCritical:
+    percent = 100;
+    break;
+  default:
+    throw std::runtime_error("Attempted to assign an unrecognized thread priority");
+  }
+  param.sched_priority = PTHREAD_MIN_PRIORITY + (percent*(PTHREAD_MAX_PRIORITY - PTHREAD_MIN_PRIORITY) + 50) / 100;
+
+  pthread_setschedparam(m_thisThread.native_handle(), policy, &param);
```
Comment (hham) [inline src/autowiring/CoreThreadMac.cpp]: m_state-> is missing in front of m_thisThread.

File: src/autowiring/CoreThreadMac.cpp [hunk 3]
```diff
@@ -11,8 +11,17 @@
...
 #include <iostream>
+#include <pthread.h>
 #include THREAD_HEADER
 
+// Missing definitions from pthreads.h
+#if !defined(PTHREAD_MIN_PRIORITY)
+#define PTHREAD_MIN_PRIORITY  0
+#endif
+#if !defined(PTHREAD_MAX_PRIORITY)
+#define PTHREAD_MAX_PRIORITY 31
```
Comment (hham) [inline src/autowiring/CoreThreadMac.cpp]: It looks like 63 is max priority in user space according to https://developer.apple.com/library/mac/documentation/Darwin/Conceptual/KernelProgramming/scheduler/scheduler.html.
Why did you put 31?

Comment (gittyupagain) [inline src/autowiring/CoreThreadMac.cpp]: This was done 20 months ago and never committed. It was an initial attempt at the problem. Those numbers probably came from Linux documentation. It is unfortunate that it isn't defined. If 63 is correct, then that should be used instead.

Comment (codemercenary) [inline src/autowiring/CoreThreadMac.cpp]: I'll make the change.

**Pull request merged**